src/Database/LSMTree/Internal/IndexCompactAcc.hs

```diff
@@ -127,6 +130,18 @@ new rfprec maxcsize =
     <*> newSTRef SNothing
     <*> newSTRef SNothing
 
+-- | We explictly pin the byte arrays, since that allows for more efficient
+-- serialisation, as the definition of 'isByteArrayPinned' changed in GHC 9.6,
+-- see <https://gitlab.haskell.org/ghc/ghc/-/issues/22255>.
+--
+-- TODO: remove this workaround once a solution exists, e.g. a new primop that
+-- allows checking for implicit pinning.
+newPinnedMVec :: Int -> ST s (VUM.MVector s Word32)
+newPinnedMVec lenWords = do
+    mba <- newPinnedByteArray (mul4 lenWords)
+    setByteArray mba 0 lenWords (0 :: Word32)
+    return (VUM.MV_Word32 (VP.MVector 0 lenWords mba))
```

Comment:
Interestingly, this doesn't make a big difference, as the index is much smaller than e.g. the entries themselves. Most of the allocation reductions in the changes to index serialisation already occur when the chunks are not pinned, in particular the big improvements for the cases with only large entries. Not quite sure where they come from.